pkg/ccl/sqlproxyccl/proxy_handler.go

```diff
@@ -420,42 +425,35 @@ func (handler *proxyHandler) handle(ctx context.Context, incomingConn *proxyConn
 
 	handler.metrics.SuccessfulConnCount.Inc(1)
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	log.Infof(ctx, "new connection")
 	connBegin := timeutil.Now()
 	defer func() {
 		log.Infof(ctx, "closing after %.2fs", timeutil.Since(connBegin).Seconds())
 	}()
 
-	// Copy all pgwire messages from frontend to backend connection until we
-	// encounter an error or shutdown signal.
-	go func() {
-		err := ConnectionCopy(crdbConn, conn)
-		select {
-		case errConnection <- err: /* error reported */
-		default: /* the channel already contains an error */
-		}
-	}()
+	// Pass ownerships of conn and crdbConn to the forwarder.
+	f := newForwarder(ctx, conn, crdbConn)
+	defer f.Close()
+
+	// Start forwarding messages.
+	if err := f.Run(); err != nil {
+		// This should not happen.
+		panic(errors.Wrap(err, "forwarder.Run failed"))
+	}
 
+	// Block until an error is received, or when the stopper starts quiescing,
+	// whichever that happens first.
 	select {
-	case err := <-errConnection:
+	case err := <-f.errChan: // From forwarder.
```

Comment:
What do you think about pushing all of the logic below authenticate into Forwarder? That allows us to merge the goroutine waiting on the errChan and the goroutine waiting on the context cancellation.

In my mind Forwarder is a representation of the connection's state machine. So it feels reasonable to make it responsible for handling the idle and error states.